### PR TITLE
Fix resource resolution in `InstructionSet.c_if`

### LINKED ISSUE #7246: `Clbit` indexing in `c_if` inconsistent with circuit indexing
### Information

- **Qiskit Terra version**: main
- **Python version**: any
- **Operating system**: any

### What is the current behavior?

If two classical registers have a non-empty intersection of clbits, it's possible to have `InstructionSet.c_if` resolve an integer index to a different value than the original `QuantumCircuit` would.

### Steps to reproduce the problem

```python
In [1]: from qiskit.circuit import QuantumCircuit, ClassicalRegister, Clbit, Qubit
   ...: clbits = [Clbit() for _ in [None]*3]
   ...: # Two overlapping classical registers.
   ...: cr1 = ClassicalRegister(bits=clbits[:2])
   ...: cr2 = ClassicalRegister(bits=clbits[1:])
   ...:
   ...: qc = QuantumCircuit([Qubit()], cr1, cr2)
   ...: qc.x(0).c_if(2, 0)
   ...: qc.measure(0, 2)
   ...:
   ...: clbit_from_c_if = qc.data[0][0].condition[0]
   ...: clbit_from_measure = qc.data[1][2][0]
   ...: clbit_from_c_if is clbit_from_measure
Out[1]: False
```

### What is the expected behavior?

They should resolve to the same clbit, because they use the same index.

### Suggested solutions

Instead of storing references to the classical registers in the instruction set, store a reference to the circuit, or a bound callable that came from the circuit to resolve the indices.

### LINKED ISSUE #7247: Chained calls to `c_if` overwrite previous `condition`
### Information

- **Qiskit Terra version**: main
- **Python version**: any
- **Operating system**: any

### What is the current behavior?

`Instruction.c_if` overwrites the previous condition, regardless of whether or not a condition existed.  This means that chained calls do not produce nested `if` statements, but instead replace the call.

### Steps to reproduce the problem

```python
In [1]: from qiskit.circuit import QuantumCircuit, ClassicalRegister, Qubit
   ...: from qiskit.circuit.library import XGate
   ...: cr = ClassicalRegister(2)
   ...:
   ...: cond1 = (cr[0], False)
   ...: cond2 = (cr[1], True)
   ...:
   ...: qc = QuantumCircuit([Qubit()], cr)
   ...: qc.x(0).c_if(*cond1).c_if(*cond2)
   ...: qc.draw()
Out[1]:
         ┌───┐
   0: ───┤ X ├────
         └─╥─┘
      ┌────╨─────┐
c0: 2/╡ c0_1 = T ╞
      └──────────┘

In [2]: XGate().c_if(*cond2) == qc.data[0][0]
Out[2]: True
```

### What is the expected behavior?

There's no longer any reference to the first condition, whereas the documentation for `c_if` says it _adds_ a condition, not overwrites it.

### Suggested solutions

Fix the documentation of `c_if`.  While a nesting behaviour would likely be more expected, there's no real need to do that since QASM 2 doesn't support it and we'd have to completely overhaul how conditions work. The new QASM 3-ish control-flow modes (#7123) are already coming, and they support arbitrary nesting.  In the long term, we can probably deprecate `c_if`.

### LINKED ISSUE #7249: Each `c_if` call linear in number of circuit `Clbit`s
### Information

- **Qiskit Terra version**: main
- **Python version**: any
- **Operating system**: any

### What is the current behavior?

Any call to `Instruction.c_if` is linear in the number of `Clbit` references in classical registers in the circuit.  This is because of some lines in `InstructionSet` that unroll all registers on every call: https://github.com/Qiskit/qiskit-terra/blob/49110f9c596fc509f189f38cfbdbb3100407bff2/qiskit/circuit/instructionset.py#L61-L64

### Steps to reproduce the problem

```python
In [1]: from qiskit.circuit import QuantumCircuit
   ...: import numpy as np
   ...: import scipy.stats
   ...:
   ...: def build(n):
   ...:     qc = QuantumCircuit(1, n)
   ...:     instructions = qc.x(0)
   ...:     def payload():
   ...:         instructions.c_if(0, 0)
   ...:     return payload
   ...:
   ...: times = []
   ...: ns = [100, 300, 1000, 3000, 10000]
   ...: for n in ns:
   ...:     payload = build(n)
   ...:     out = %timeit -o payload()
   ...:     times.append(out.average)
   ...: scipy.stats.linregress(np.log(ns), np.log(times)).slope
10.8 µs ± 101 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
28.2 µs ± 922 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
94 µs ± 640 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
276 µs ± 5.04 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
919 µs ± 7.14 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Out[1]: 0.9710414298958975
```

### What is the expected behavior?

This should be a constant-time operation.

### Suggested solutions

This can easily be fixed by the same methods as #7246.

### LINKED ISSUE #7250: `c_if` fails when using indexing if there are no classical registers in the circuit
### Information

- **Qiskit Terra version**: main
- **Python version**: any
- **Operating system**: any

### What is the current behavior?

All `c_if` statements immediately fail if there are no classical registers defined in the circuit, only loose bits, and one tries to specify the condition with an index.

### Steps to reproduce the problem

```python
In [1]: from qiskit.circuit import QuantumCircuit, Clbit, Qubit
   ...:
   ...: qc = QuantumCircuit([Qubit(), Clbit()])
   ...: qc.x(0).c_if(0, 0)
---------------------------------------------------------------------------
UnboundLocalError                         Traceback (most recent call last)
<ipython-input-1-efd3b9b41798> in <module>
      2
      3 qc = QuantumCircuit([Qubit(), Clbit()])
----> 4 qc.x(0).c_if(0, 0)

~/code/qiskit/terra/qiskit/circuit/instructionset.py in c_if(self, classical, val)
     65         if isinstance(classical, (Clbit, int)):
     66             if isinstance(classical, int):
---> 67                 classical = circuit_clbits[classical]
     68             for creg in self.circuit_cregs:
     69                 if creg[:] == [classical]:

UnboundLocalError: local variable 'circuit_clbits' referenced before assignment
```

### What is the expected behavior?

There should be no error, and it should select the correct bit as a condition.

### Suggested solutions

A similar fix to #7246 will work here as well.

### PR DESCRIPTION
**Pull request by jakelishman**
### "Summary"

This completely overhauls `InstructionSet.c_if` to use the containing
`QuantumCircuit` instance (if there is one) to resolve requests for
classical resources.  This has several benefits:

  1. The meaning of a integer is guaranteed to be the same between
     instances of `QuantumCircuit` and `InstructionSet`.

  2. There's no need for code duplication between the two.

  3. With the resolution tied to pre-existing instance methods or
     functions now, there's no need for each `InstructionSet` instance
     to build up its own resolution graph over registers, which is a
     speed boost.

  4. It makes it possible for resolvers in the future to reject certain
     requests for resources, if they cannot be satisfied.  This can
     easily be used by scoping constructs, to forbid access to extra
     resource once the scope has closed, since `InstructionSet.c_if` can
     mutate the contents after the fact.

Since `InstructionSet` is technically part of the public API, albeit not
one we expect people to actually be using, this deprecates the
`circuit_cregs` keyword argument, and maintains a compatibility layer so
existing calls will continue to work until the deprecation period
expires.  All internal usages of `InstructionSet` use the new form.

There remains an unfortunate linear complexity in the new resource
checker, when seeing if a `ClassicalRegister` target is valid.  This is
because there is currently no set-like interface to the registers stored
in any given circuit, unlike the bits.  `QuantumCircuit` exposes its
lists of registers publically, so adding this map safely is non-trivial,
as it is simple for a user or subclasses to cause any new map to fall
out of sync with the list.  This commit does not attempt to fix this,
but simply suffers the linear penalty for now; we expect there to be
only a few registers, so it should not be too onerous.

The previous form of `InstructionSet.c_if` had a hack to support a
limitation of `disassemble`.  Previously, any single-bit conditions
where the bit was part of a register of size one would get automatically
promoted to be conditional on the register, not the bit.  With
`QasmQobj`, single-bit conditions are indistinguishable from conditions
on size-one registers, so `InstructionSet.c_if` promoted them all to
register comparisons to try and make `disassemble(assemble(qc))` a
complete round-trip in those cases.  Since `QasmQobj` cannot cope with
loose classical bits or overlapping registers, anyway, this was at best
a stop-gap measure, as not all circuits can be round-tripped already.
This commit removes the auto-promotion behaviour from
`InstructionSet.c_if`, but adds the logic to `disassemble` instead,
noting in the documentation that not all circuits will produce exactly
equivalent output when round-tripped.  This is to make the "type system"
of classical conditions strict, which significantly simplifies the
resource management.

### Details and comments

This PR isn't as big as it looks - it's 90% tests and documentation.

The idea of a "resource resolver" callback is better for consistency, but I also designed it to do double duty.  The control-flow circuit builder interface would struggle to work safely with `c_if`, because `c_if` mutates instructions after they've passed through `QuantumCircuit.append`, which is where the builders hook in to build scopes and manage resources, but having `c_if` call back to a resource manager lets them work safely again (and also ensures that you can't mess with instructions once the scope has closed).

- Fix #7246 by hooking into the `QuantumCircuit` instance to resolve resources, not just the registers.
- Fix #7247 by updating the documentation of `InstructionSet.c_if` and `Instruction.c_if`.
- Fix #7249 by indexing into `QuantumCircuit`'s bit lists, not constructing a new one each time.
- Fix #7250 by hooking into the `QuantumCircuit` instance to resolve resources, not just the registers.
- See #6989 for the original implementation of the `disassemble`/`c_if` size-one register promotion.
- See #7133 for more details on the scaling of look-ups of registers.

### CODE REVIEW

**Comment by jakelishman** [comment]
On hold until #7123 merges, because while there won't be any merge conflicts, logically this PR will want to make a couple of changes to some error checking routines in that PR once it's in.  The code here is still good for immediate review, though.

**Comment by jakelishman** [comment]
Updated to account for the extra changes in #7123, so this is no longer on hold.

File: releasenotes/notes/fix-instruction-c_if-3334bc8bcc38a327.yaml
Comment (mtreinish) [inline releasenotes/notes/fix-instruction-c_if-3334bc8bcc38a327.yaml:15]: Do you want to add issue numbers and links to the issues here?

File: releasenotes/notes/fix-instruction-c_if-3334bc8bcc38a327.yaml
Comment (mtreinish) [inline releasenotes/notes/fix-instruction-c_if-3334bc8bcc38a327.yaml:5]: I think having an example of how this is done would be useful. I agree it's not really user facing, but reading this with my pretend naive user eye I don't know what this would look like.

File: qiskit/circuit/instructionset.py
Comment (mtreinish) [inline qiskit/circuit/instructionset.py:170]: Maybe expand on this a bit and say it will happen silently too. I just don't want people to be surprised that if you do `qc.x(0).c_if(qr, 2).c_if(qr[0], 1)` that there is only a condition on the first bit.

File: qiskit/circuit/instructionset.py
Comment (mtreinish) [inline qiskit/circuit/instructionset.py:195]: It is a bit weird if we don't allow chaining here that we return `self` like this.

**Review by mtreinish — COMMENTED** [review]
Overall this LGTM, I especially like the documentation improvements. Just a few inline comments and questions but mostly just minor things and existing quirks with this functionality.

I also don't think the linear scan for registers will be that bad in practice. It's not like there are circuits out there with thousands of registers, I'd say the typical use case is one classical register for all the measurement bits. In the worst case it'll be a handful but like never more than 10, probably nothing where we should have scaling issues.

File: test/python/circuit/test_instructions.py
Comment (kdk) [inline test/python/circuit/test_instructions.py:521]: Is using `assertRaisesRegex` a possibility here (just to make sure we're hitting the right `CircuitError`)?

File: qiskit/circuit/instructionset.py
Comment (kdk) [inline qiskit/circuit/instructionset.py:120]: Why would the `resource_resolver` need (or want) to assume a call implies the resource will be used?

**Review by kdk — COMMENTED** [review]
Few small questions, LGTM otherwise.

File: qiskit/circuit/instructionset.py
Comment (jakelishman) [reply qiskit/circuit/instructionset.py:120]: It's 100% for the control-flow builder interface.  The call to `InstructionSet.c_if` occurs after the call to `QuantumCircuit.append`, which is where the scoping constructs hook into, so normally they wouldn't be able to track any additional classical resources that are inserted by a `c_if` call.  With this set-up, once you enter a scope, the resolver for the `InstructionSet` now comes from the scope instead, which lets it track injected resources.

Perhaps the name of the function is wrong here - it's not _just_ a resolver, it's actually resolves, verifies and tracks resources that are given to it.

File: releasenotes/notes/fix-instruction-c_if-3334bc8bcc38a327.yaml
Comment (jakelishman) [reply releasenotes/notes/fix-instruction-c_if-3334bc8bcc38a327.yaml:4]: The leading `.` means search the symbol table for a symbol with this suffix, and then display only the suffix.  In this case it resolves fully to `qiskit.circuit.InstructionSet`.  For dot access, the search order goes from "nearest" to "furthest" roughly, so if I reference `.InstructionSet` in `qiskit.a.b`, it'll search
1. `qiskit.a.b.InstructionSet`
2. `qiskit.a.InstructionSet`
3. `qiskit.InstructionSet`
4. `InstructionSet`
5. _anything_`.InstructionSet`
I think it's meant to warn if there are duplicate candidates for resolution, but I usually only use the shorthand when I'm fairly sure there's only one thing it could match anyway, because otherwise it's likely to be confusing in the text as well.

`:obj:` is magic and resolves to whatever role the name is first found in, so it can match `:class:`, `:meth:`, `:attr:`, etc, but if the name exists in more than one role, you can't generally predict which you'll get.  In this case, there's only one `InstructionSet`, so it'll

File: qiskit/assembler/disassemble.py
Comment (jakelishman) [reply qiskit/assembler/disassemble.py:46]: `` :obj:`.x` `` only displays what's after the `.` (I think), and in module/class `a.b.C` it has the search path:
1. `a.b.C.x`
2. `a.b.x`
3. `a.x`
4. `x`
5. _any prefix_`.x`
Sphinx is meant to warn if there are duplicate names that it could resolve to, but I've never actually tried that myself - if I know there are potentially duplicate names, I tend to use slightly longer forms because if it's confusing for me/Sphinx, it'll probably be confusing for the reader too.  For example, in the release notes I used `` :meth:`.InstructionSet.c_if` `` to avoid conflict with `Instruction.c_if`, and that displays like this:

Screenshot 2021-11-16 at 10 05 15

File: qiskit/circuit/instructionset.py
Comment (jakelishman) [reply qiskit/circuit/instructionset.py:195]: Yeah, seems slightly odd. Perhaps we were thinking of adding more methods to `Instruction`/`InstructionSet` so you may want to do multiple things at once.

File: releasenotes/notes/fix-instruction-c_if-3334bc8bcc38a327.yaml
Comment (jakelishman) [reply releasenotes/notes/fix-instruction-c_if-3334bc8bcc38a327.yaml]: Good idea, done in 2c608d212.  The last two don't have issues associated with them, so I've linked the PR instead.

File: releasenotes/notes/fix-instruction-c_if-3334bc8bcc38a327.yaml
Comment (jakelishman) [reply releasenotes/notes/fix-instruction-c_if-3334bc8bcc38a327.yaml:5]: Sounds good, done in 2c608d212.

File: qiskit/circuit/instructionset.py
Comment (jakelishman) [reply qiskit/circuit/instructionset.py]: Sure - this isn't changed behaviour, it's just adding a documentation note on something that was already happening silently, so noisy is better.  With the current way `Instruction.condition` works, we can't really support compound conditions at all, but when we get a bit more QASM3-ish support, we'll need to overhaul that field/remove it in favour of `IfElseOp` and we can have a proper boolean expression in some form.

Done in e0b3700ae.

File: releasenotes/notes/fix-instruction-c_if-3334bc8bcc38a327.yaml
Comment (jakelishman) [reply releasenotes/notes/fix-instruction-c_if-3334bc8bcc38a327.yaml]: Fixed in 2c608d212, thanks.

File: test/python/circuit/test_instructions.py
Comment (jakelishman) [reply test/python/circuit/test_instructions.py]: Oh crumbs - I was just writing the builder interface with all `assertRaisesRegex` and thinking "I'm sure I did this incorrectly only recently".  This must have been what I was thinking of!

Fixed in 9f87c4adc.

File: releasenotes/notes/fix-instruction-c_if-3334bc8bcc38a327.yaml
Comment (kdk) [reply releasenotes/notes/fix-instruction-c_if-3334bc8bcc38a327.yaml:4]: Awesome, that's super handy to know!

File: qiskit/circuit/instructionset.py
Comment (kdk) [reply qiskit/circuit/instructionset.py:120]: Got it, okay. Maybe in the context of the builder PR, a better name might come to mind, but I think for now this is good as is.

**Review by kdk — DISMISSED** [review]
Thanks for this updates, this LGTM.

File: qiskit/circuit/instructionset.py
Comment (jakelishman) [reply qiskit/circuit/instructionset.py:120]: Perhaps `request_resource` instead of `resolve_resource`?  That has the right sort of implications in the name, I think.

Comment (jakelishman) [reply qiskit/circuit/instructionset.py]: Changed to `resource_requester` in f8ce816dd.

**Review by kdk — APPROVED** [review]

**Pull request merged**
